typed-racket-lib/typed-racket/static-contracts/optimize.rkt

```diff
@@ -117,23 +117,106 @@
     [(arr/sc: args rest (list (any/sc:) ...))
      (arr/sc args rest #f)]
     [(none/sc:) any/sc]
-    [(app sc-terminal-kind 'flat) any/sc]
+    [(or/sc: (? flat-terminal-kind?) ...) any/sc]
+    [(? flat-terminal-kind?) any/sc]
     [else sc]))
 
+(define (flat-terminal-kind? sc)
+  (eq? 'flat (sc-terminal-kind sc)))
 
+;; The side of a static contract describes the source of the values that
+;;  the contract needs to check.
+;; - 'positive : values exported by the server module
+;; - 'negative : values imported from a client module
+;; - 'both     : values from both server & client
+(define (side? v)
+  (memq v '(positive negative both)))
+
+;; A _weak side_ is a side that is currently unsafe to optimize
+;; Example:
+;;  when optimizing an `(or/sc scs ...)` on the 'positive side,
+;;  each of the `scs` should be optimized on the '(weak positive) side,
+;;  and their sub-contracts --- if any --- may be optimized on the 'positive side
+(define (weak-side? x)
+  (match x
+   [(list 'weak (? side?))
+    #true]
+   [_
+    #false]))
+
+(define (strengthen-side side)
+  (if (weak-side? side)
+    (second side)
+    side))
+
+(define (weaken-side side)
+  (if (weak-side? side)
+    side
+    `(weak ,side)))
 
 (define (invert-side v)
-  (case v
-    [(positive) 'negative]
-    [(negative) 'positive]
-    [(both) 'both]))
+  (if (weak-side? v)
+    (weaken-side (invert-side v))
+    (case v
+      [(positive) 'negative]
+      [(negative) 'positive]
+      [(both) 'both])))
 
 (define (combine-variance side var)
   (case var
     [(covariant) side]
     [(contravariant) (invert-side side)]
     [(invariant) 'both]))
 
+;; update-side : sc? weak-side? -> weak-side?
+;; Change the current side to something safe & strong-as-possible
+;;  for optimizing the sub-contracts of the given `sc`.
+(define (update-side sc side)
+  (match sc
+   [(or/sc: scs ...)
+    #:when (not (andmap flat-terminal-kind? scs))
+    (weaken-side side)]
+   [(? guarded-sc?)
+    (strengthen-side side)]
+   [_
+    ;; Keep same side by default.
+    ;; This is precisely safe for "unguarded" static contracts like and/sc
+    ;;  and conservatively safe for everything else.
+    side]))
+
+;; guarded-sc? : sc? -> boolean?
+;; Returns #true if the given static contract represents a type with a "real"
+;;  type constructor. E.g. list/sc is "real" and or/sc is not.
+(define (guarded-sc? sc)
+  (match sc
```

Comment:
I was swimming through this code and had the thought -- would it be easier to recognize the type constructors that are _not_ real (i.e. or/sc, and/sc... is that it? maybe 1 or two more?) instead of all the ones that are real?

Maybe there's a reason why that would be problematic, but I figured I would ask just in case.

Reply:
(also -- I know this PR was merged a while ago, but I wasn't sure where to ask the question about this line... ¯\\\_(ツ)\_/¯)

Reply:
That would be shorter, but I picked this because its safer. If `guarded-sc?` leaves something out, then its only a missed optimization and not a soundness hole.